### PR TITLE
JSON parses all numbers as floats, this causes increased complexity downstream, converting to int64 here is best. Added a benchmark test and verified it's negligable overhead.

### PR DESCRIPTION
**Pull request by mdubbyap**

### CODE REVIEW

**Comment by mdubbyap** [comment]
updated to use type conversion to and back and tests that if the compiler is going to optimize them out will fail

File: protocol/collectd/collectd.go [hunk 1]
```diff
@@ -75,7 +75,13 @@ func NewDatapoint(point *JSONWriteFormat, index uint, defaultDimensions map[stri
 	addIfNotNullOrEmpty(dimensions, "dsname", !usedDsName, dsname)
 
 	timestamp := time.Unix(0, int64(float64(time.Second)**point.Time))
-	return datapoint.New(metricName, dimensions, datapoint.NewFloatValue(*val), metricType, timestamp)
+	var value datapoint.Value
+	if *val == float64(int64(*val)) {
```
Comment (mpetazzoni) [inline protocol/collectd/collectd.go]: You might save a bit by keeping around the int value (although that probably gets optimized?):

```go
intVal := int64(*val)
if *val == float64(intVal) {
    value = datapoint.NewIntValue(intVal)
} else {
    value = datapoint.newFloatValue(*val)
}
```

File: protocol/signalfx/signalfx.go [hunk 2]
```diff
@@ -38,6 +38,9 @@ type ValueToSend interface {
 func ValueToValue(v ValueToSend) (datapoint.Value, error) {
 	f, ok := v.(float64)
 	if ok {
+		if f == float64(int64(f)) {
+			return datapoint.NewIntValue(int64(f)), nil
```
Comment (mpetazzoni) [inline protocol/signalfx/signalfx.go]: Same

File: protocol/collectd/collectd.go (hunk 1 above)
Comment (mdubbyap) [inline protocol/collectd/collectd.go]: good eye. I tried that out in the benchmark test and had no effect on performance or allocations, so it gets optimized out and i felt the original was more readable.

**Pull request merged**